### ISSUE TITLE
Extend prefix usage in N3Writer

### ISSUE THREAD

**Issue by rubensworks** [RubenVerborgh/N3.js#55]
Apparently, N3 does not support the complete set of allowed characters in prefix handling.
For example if we have the URL `http://example.org/graphs#_0`, we expect it to shorten it to `g:_0` if we have the prefix `g: <http://example.org/graphs#>`.
Unfortunately, [it does not do this](https://github.com/RubenVerborgh/N3.js/blob/master/lib/N3Writer.js#L224), while the [Turtle spec](http://www.w3.org/TeamSubmission/turtle/#name) says that it should, including support for a bunch of other special characters.

In my [fork](https://github.com/rubensworks/N3.js/commit/01403f47fe48aeb78dc8fb5459430be08435483b), I added full support for this with the purpose of sending a PR, but while executing the tests I ran into an interesting issue.

The Turtle spec shows that prefixed entities like `g:123` are not allowed, they should start with a [character or something similar](http://www.w3.org/TeamSubmission/turtle/#nameStartChar), and there are unit tests in this repo to confirm this behaviour.
Then I discovered that the [nameStartChar](http://www.w3.org/TeamSubmission/turtle/#nameStartChar) also allows a character from the range `[#x10000-#xEFFFF]`, and apparently this range includes a lot of characters.

If you would open up a node console and execute

    /[\u10000-\uEFFFF]/.test("1")

You would see that it accepts all numbers (including a lot of other special characters as well).
So I am not sure if this specific range is intended to include numbers, if so, this would require some changes in the unit tests.
I assume this is a mistake in the spec, because for example the term [nameChar](http://www.w3.org/TeamSubmission/turtle/#nameChar) explicitly accepts the range `[0-9]`, even though it was already accepted by the range `[#x10000-#xEFFFF]`.

**Comment by RubenVerborgh** [comment]
Thanks, will investigate!

**Comment by RubenVerborgh** [comment]
Alright, so there's two parts here.

The first is _generation_. It's true that [the regex in N3Writer.js](https://github.com/RubenVerborgh/N3.js/blob/v0.4.5/lib/N3Writer.js#L224) does not abbreviate all IRIs that are allowed to be abbreviated. However, this is not an error against the spec: while the Turtle spec says that all parsers should be able to deal with such abbreviations, it is not mandatory that writers actually use all possible abbreviations at every possible instance. (In other words: every valid N-Triples file is also a valid Turtle file, even if you prepend a whole list of prefixes to it.)

So why would we not abbreviate in all possible cases then? Performance. Executing more complicated regex might take a lot more time, so if it is not “obvious” that we can abbreviate, we simply do not abbreviate. However, I'd be open to see just how complex we can make the regex without a performance drop, but this would require a decent benchmark.

So in short: no violation of the spec, but possible room for optimization.

The second part is about the Turtle spec. Let me take some time to investigate that as well.

**Comment by RubenVerborgh** [comment]
For the second part, the explanation is in [JavaScript’s internal character encoding: UCS-2 or UTF-16?](https://mathiasbynens.be/notes/javascript-encoding).

Consider the following commands:
```
> /[\u10000-\uEFFFF]/.test("1")
true
> "1".charCodeAt(0).toString(16)
'31'
> /\u0031/.test("1")
true
> "\u10000-\uEFFFF".substr(1,1)
'0'
```
In other words:
- the actual unicode-escape for the character `"1"` is `"\u0031"`
- your regex `/[\u10000-\uEFFFF]/`. is actually equivalent to `/[0-\uEFFF\u1000F]/` and thus `/[0-\uEFFF]|\u1000|F/`, and `/[0-\uEFFF]/`indeed includes `"1"`.

**Comment by RubenVerborgh** [comment]
So I can accept your pull request:
- if performance is shown to remain OK
- if the [\u10000-\uEFFFF] part of the regex is dropped or fixed (for help, see #3)

**Comment by rubensworks** [comment]
Yes, lowered performance of the regex evaluation is indeed something I was also thinking about with those enlarged expressions, I can imagine many of these ranges lowering the speed of the evaluation, so  a simple timing experiment should provide some insights into this.

An alternative would be to handle the regex in such a way that developers can *extend* it to support more special characters, like for example for the character `à`, when it is known what range of characters is expected.

For the second part; it indeed seemed more likely that it was a mistake on my end and not with the Turtle spec. I'll look into it.

**Comment by RubenVerborgh** [comment]
Closing this for now until we have performance tests.
